source/MaterialXTest/RenderUtil.h

```diff
@@ -0,0 +1,292 @@
+    std::string _label;
+    std::chrono::time_point<std::chrono::system_clock> _startTime;
+};
+
+
+// Per language profile times
+//
+class LanguageProfileTimes
+{
+  public:
+    void print(const std::string& label, std::ostream& output) const
+    {
+        output << label << std::endl;
+        output << "\tTotal: " << totalTime << " seconds" << std::endl;;
+        output << "\tSetup: " << setupTime << " seconds" << std::endl;;
+        output << "\tTransparency: " << transparencyTime << " seconds" << std::endl;;
+        output << "\tGeneration: " << generationTime << " seconds" << std::endl;;
+        output << "\tCompile: " << compileTime << " seconds" << std::endl;
+        output << "\tRender: " << renderTime << " seconds" << std::endl;
+        output << "\tI/O: " << ioTime << " seconds" << std::endl;
+        output << "\tImage save: " << imageSaveTime << " seconds" << std::endl;
+    }
+    double totalTime = 0.0;
+    double setupTime = 0.0;
+    double transparencyTime = 0.0;
+    double generationTime = 0.0;
+    double compileTime = 0.0;
+    double renderTime = 0.0;
+    double ioTime = 0.0;
+    double imageSaveTime = 0.0;
+};
+
+// Render validation profiling structure
+//
+class RenderProfileTimes
+{
+  public:
+    void print(std::ostream& output) const
+    {
+        output << "Overall time: " << languageTimes.totalTime << " seconds" << std::endl;
+        output << "\tI/O time: " << ioTime << " seconds" << std::endl;
+        output << "\tValidation time: " << validateTime << " seconds" << std::endl;
+        output << "\tRenderable search time: " << renderableSearchTime << " seconds" << std::endl;
+
+        languageTimes.print("Profile Times:", output);
+
+        output << "Elements tested: " << elementsTested << std::endl;
+    }
+
+    LanguageProfileTimes languageTimes;
+    double totalTime = 0;
+    double ioTime = 0.0;
+    double validateTime = 0.0;
+    double renderableSearchTime = 0.0;
+    unsigned int elementsTested = 0;
+};
+
+// Base class used for performing compilation and render tests for a given
+// shading language and target.
+//
+class ShaderRenderTester
+{
+  public:
+    ShaderRenderTester() {};
+    virtual ~ShaderRenderTester() {};
+    bool validate();
+
+  protected:
+    // The shading language / target being tested
+    virtual const std::string& languageTargetString() = 0;
+
+    // Check if testing should be performed based in input options
+    virtual bool runTest(const RenderUtil::RenderTestOptions& testOptions) const = 0;
+
+    // Load any addition libraries requird by the generator
+    virtual void loadLibraries(mx::DocumentPtr /*dependLib*/,
+                               RenderUtil::RenderTestOptions& /*options*/) {};
+
+    //
+    // Code generation methods
+    //
+    // Create the appropirate code generator for the language/target
+    virtual void createShaderGenerator() = 0;
+    // Register any addition source code paths used by the generator
```

Comment:
Typo: additional

Reply:
Nitpick: for code convention I think we normally use a newline before each method.